samples/drivers/adc/src/main.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020 Libre Solar Technologies GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/printk.h>
+#include <drivers/adc.h>
+
+#if !DT_NODE_EXISTS(DT_PATH(zephyr_user)) || \
+	!DT_NODE_HAS_PROP(DT_PATH(zephyr_user), io_channels)
+#error "No suitable devicetree overlay specified"
+#endif
+
+#define ADC_NUM_CHANNELS	DT_PROP_LEN(DT_PATH(zephyr_user), io_channels)
+
+#if ADC_NUM_CHANNELS > 2
+#error "Currently only 1 or 2 channels supported in this sample"
+#endif
+
+/* Check that the same ADC is used for first and second phandle */
+#if ADC_NUM_CHANNELS == 2 && \
+	DT_PHANDLE_BY_IDX(DT_PATH(zephyr_user), io_channels, 0) != \
+	DT_PHANDLE_BY_IDX(DT_PATH(zephyr_user), io_channels, 1)
+#error "Channels have to use the same ADC."
+#endif
+
+#define ADC_NODE		DT_PHANDLE(DT_PATH(zephyr_user), io_channels)
+
+/* Check if the specified ADC is enabled */
+#if DT_NODE_HAS_STATUS(ADC_NODE, okay)
+#define ADC_DEVICE_NAME		DT_LABEL(ADC_NODE)
+#else
+#error "ADC disabled or not available for this board"
+#endif
+
+/* Common settings supported by most ADCs */
+#define ADC_RESOLUTION		12
+#define ADC_GAIN		ADC_GAIN_1
+#define ADC_REFERENCE		ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+
+/* Get the numbers of up to two channels */
+static uint8_t channel_ids[ADC_NUM_CHANNELS] = {
+	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 0),
+#if ADC_NUM_CHANNELS == 2
+	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 1)
+#endif
+};
+
+static int16_t sample_buffer[ADC_NUM_CHANNELS];
+
+struct adc_channel_cfg channel_cfg = {
+	.gain = ADC_GAIN,
+	.reference = ADC_REFERENCE,
+	.acquisition_time = ADC_ACQUISITION_TIME,
+	/* channel ID will be overwritten below */
+	.channel_id = 0,
+	.differential = 0
+};
+
+struct adc_sequence sequence = {
+	/* individual channels will be added below */
+	.channels    = 0,
+	.buffer      = sample_buffer,
+	/* buffer size in bytes, not number of samples */
+	.buffer_size = sizeof(sample_buffer),
+	.resolution  = ADC_RESOLUTION,
+};
+
+void main(void)
+{
+	int err;
+	const struct device *dev_adc = device_get_binding(ADC_DEVICE_NAME);
```

Comment:
I think it would be nicer to start using DEVICE_DT_GET in new samples. I think that style is easier to read:

```
const struct device *dev_adc = DEVICE_DT_GET(ADC_NODE);

if (!device_is_ready(dev_adc)) {
	printk("ADC device not found\n");
	return;
}
```

Then you don't need to define `ADC_DEVICE_NAME`.

This is not blocking, just a suggestion.